muffnn/autoencoder/autoencoder.py

```diff
@@ -0,0 +1,682 @@
+"""
+Autoencoder in scikit-learn style with TensorFlow
+"""
+import logging
+import re
+import warnings
+
+import numpy as np
+import tensorflow as tf
+import scipy.sparse as sp
+
+from sklearn.base import TransformerMixin, BaseEstimator
+from sklearn.utils import check_array, check_random_state
+from sklearn.utils.validation import NotFittedError
+
+from ..core import TFPicklingBase, affine
+
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class Autoencoder(TFPicklingBase, TransformerMixin, BaseEstimator):
+    """
+    An autoencoder using TensorFlow.
+
+    Parameters
+    ----------
+    hidden_units : tuple or list, optional
+        A list of integers indicating the number of hidden layers and their
+        sizes. The list is reflected about the last element to form the
+        autoencoder. For example, if hidden_units is (256, 32, 16) then
+        the autoencoder has layers
+
+            [input_layer, 256, 32, 16, 32, 256, input_layer]
+
+    batch_size : int, optional
+        The batch size for learning and prediction. If there are fewer
+        examples than the batch size during fitting, then the the number of
+        examples will be used instead.
+    n_epochs : int, optional
+        The number of epochs (iterations through the training data) when
+        fitting.
+    dropout : float or None, optional
+        The dropout probability. If None, then dropout will not be used.
+        Note that dropout is not applied to the input layer.
+    activation : (callable, callable), optional
+        The activation functions for
+
+            (the bulk of the autoencoder and encoding layer, output layer).
+
+        See `tensorflow.nn` for various options.
+    random_state: int, RandomState instance or None, optional
+        If int, the random number generator seed. If RandomState instance,
+        the random number generator itself. If None, then `np.random` will be
+        used.
+    learning_rate : float, optional
+        Learning rate for Adam.
+    metric : string, optional
```

Comment:
i think i understand your approach here, and i think it makes sense, but it seems like maybe there's a better way to expose the options.  how about this?

```
"""
An autoencoder using TensorFlow for discrete, continuous, or mixed data.

Parameters
----------
....
hidden_activation : callable, optional
    The activation function for the hidden layers.
    See `tensorflow.nn` for various options.
output_activation : callable, optional
    The activation function for the output layer.
    See `tensorflow.nn` for various options.
...
metric : string, optional
    Default metric for the autoencoder. Options are
        'mse' - mean square error
        'cross-entropy' - cross-entropy
    Note that this will be overridden for columns specified by `binary_indices`
    or `categorical_indices`
binary_indices : array-like, shape (n_binary,), optional
    Array of indices for which `tf.nn.sigmoid` will be used for the
    output layer activation and cross-entropy will be used in the loss
    function.
categorical_indices : array-like, shape (n_categorical, 2), optional
    An array where each row specifies a range of indices for a categorical
    variable that has been expanded to multiple indices (e.g., 1-hot encoded).
    The first column contains start indices. The second column contains
    lengths. For each range of indices, the output layer will use
    `tensorflow.nn.softmax` activation and the loss function will use
    cross-entropy.
...
"""
```

also, I don't think cross-entropy should force a sigmoid output layer.  One might want to use either softmax or sigmoid activation with cross-entropy, I think, depending on the problem.